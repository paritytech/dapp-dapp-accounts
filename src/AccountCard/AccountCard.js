// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { bonds } from 'oo7-parity';

// TODO Wait for PR#18 on parity-reactive-ui for dependency
import EtherBalance from '../vendor/EtherBalance';

import IdentityIcon from '@parity/ui/lib/IdentityIcon';
import { IdentityName } from '@parity/ui/lib/IdentityName/identityName';
import Button from 'semantic-ui-react/dist/commonjs/elements/Button';
import Image from 'semantic-ui-react/dist/commonjs/elements/Image';
import Label from 'semantic-ui-react/dist/commonjs/elements/Label';
import Card from 'semantic-ui-react/dist/commonjs/views/Card';

import styles from './AccountCard.css';

class AccountCard extends PureComponent {
  render() {
    const {
      account,
      address,
      isDefault,
      isVisible,
      onMakeDefault,
      onToggleVisible
    } = this.props;

    return (
      <Card
        className={[styles.accountCard, !isVisible ? styles.hidden : ''].join(
          ' '
        )}
        onClick={isVisible ? null : () => onToggleVisible(address)}
      >
        <Card.Content>
          <Image
            as={IdentityIcon}
            floated="right"
            size="mini"
            address={address}
          />
          <Card.Header>
            <IdentityName address={address} name={account.name} unknown />
          </Card.Header>
          <Card.Meta>{address}</Card.Meta>
          <Card.Description>
            {account.meta.description && <p>{account.meta.description}</p>}
            <p>
              <FormattedMessage
                id="dapp.accounts.balance"
                defaultMessage="Balance"
              />: <EtherBalance balance={bonds.balance(address)} />
            </p>
          </Card.Description>
        </Card.Content>
        {!isDefault && (
          <Card.Content extra textAlign="center">
            {isVisible ? (
              <Button.Group widths="2">
                <Button basic primary onClick={() => onMakeDefault(address)}>
                  Make Default
                </Button>
                <Button
                  basic
                  color="grey"
                  onClick={() => onToggleVisible(address)}
                >
                  Hide
                </Button>
              </Button.Group>
            ) : (
              <Label>
                <FormattedMessage
                  id="dapp.accounts.hidden"
                  defaultMessage="Hidden"
                />
              </Label>
            )}
          </Card.Content>
        )}
      </Card>
    );
  }
}

export default AccountCard;

AccountCard.propTypes = {
  account: PropTypes.object.isRequired,
  address: PropTypes.string.isRequired,
  isDefault: PropTypes.bool,
  isVisible: PropTypes.bool.isRequired,
  onMakeDefault: PropTypes.func,
  onToggleVisible: PropTypes.func
};
