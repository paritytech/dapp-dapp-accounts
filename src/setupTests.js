// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import React from 'react';
import { IntlProvider, intlShape } from 'react-intl';
import { configure, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// Configure enzyme with React 16
configure({ adapter: new Adapter() });

/**
 * Components using the react-intl module require access to the intl context.
 * This is not available when mounting single components in Enzyme.
 * These helper functions aim to address that and wrap a valid,
 * English-locale intl context around them.
 */

// Create the IntlProvider to retrieve context for wrapping around.
const intlProvider = new IntlProvider({ locale: 'en' }, {});
const { intl } = intlProvider.getChildContext();

/**
 * When using React-Intl `injectIntl` on components, props.intl is required.
 */
const nodeWithIntlProp = node => {
  return React.cloneElement(node, { intl });
};

export const shallowWithIntl = (node, { context } = {}) => {
  return shallow(nodeWithIntlProp(node), {
    context: Object.assign({}, context, { intl })
  });
};

export const mountWithIntl = (node, { context, childContextTypes } = {}) => {
  return mount(nodeWithIntlProp(node), {
    context: Object.assign({}, context, { intl }),
    childContextTypes: Object.assign({}, { intl: intlShape }, childContextTypes)
  });
};
