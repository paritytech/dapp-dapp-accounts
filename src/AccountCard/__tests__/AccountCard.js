// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import React from 'react';
import { shallowToJson } from 'enzyme-to-json';
import Button from 'semantic-ui-react/dist/commonjs/elements/Button';

import { shallowWithIntl } from '../../setupTests';
import AccountCard from '../AccountCard';

// Mock props
const props = {
  address: '123',
  account: {
    name: 'Foo',
    meta: {
      description: 'Bar'
    }
  },
  isVisible: true,
  onMakeDefault: () => {},
  onToggleVisible: () => {}
};

// Mock bond
jest.mock('oo7-parity', () => ({
  bonds: {
    balance: address => `Bond of ${address}`
  }
}));

test('should render correctly in when isDefault', () => {
  const component = shallowWithIntl(<AccountCard {...props} isDefault />);

  expect(shallowToJson(component)).toMatchSnapshot();
});

test('should render correctly in when hidden', () => {
  const component = shallowWithIntl(
    <AccountCard {...props} isVisible={false} />
  );

  expect(shallowToJson(component)).toMatchSnapshot();
});

test('should handle onMakeDefault click', () => {
  const onMakeDefault = jest.fn();

  const component = shallowWithIntl(
    <AccountCard {...props} onMakeDefault={onMakeDefault} />
  );

  component
    .find(Button)
    .first()
    .simulate('click');
  expect(onMakeDefault).toHaveBeenCalledWith('123');
});

test('should handle onToggleVisibility click when visible', () => {
  const onToggleVisible = jest.fn();

  const component = shallowWithIntl(
    <AccountCard {...props} onToggleVisible={onToggleVisible} />
  );

  component
    .find(Button)
    .last()
    .simulate('click');
  expect(onToggleVisible).toHaveBeenCalledWith('123');
});

test('should handle onToggleVisibility click when hidden', () => {
  const onToggleVisible = jest.fn();

  const component = shallowWithIntl(
    <AccountCard
      {...props}
      onToggleVisible={onToggleVisible}
      isVisible={false}
    />
  );

  component.simulate('click');
  expect(onToggleVisible).toHaveBeenCalledWith('123');
});
