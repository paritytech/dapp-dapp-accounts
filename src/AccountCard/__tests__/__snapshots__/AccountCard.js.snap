// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly in when hidden 1`] = `
<Card
  className=" "
  onClick={[Function]}
>
  <CardContent>
    <Image
      address="123"
      as={[Function]}
      floated="right"
      size="mini"
      ui={true}
    />
    <CardHeader>
      <IdentityName
        address="123"
        name="Foo"
        unknown={true}
      />
    </CardHeader>
    <CardMeta>
      123
    </CardMeta>
    <CardDescription>
      <p>
        Bar
      </p>
      <p>
        <FormattedMessage
          defaultMessage="Balance"
          id="dapp.accounts.balance"
          values={Object {}}
        />
        : 
        <EtherBalance
          balance="Bond of 123"
        />
      </p>
    </CardDescription>
  </CardContent>
  <CardContent
    extra={true}
    textAlign="center"
  >
    <Label>
      <FormattedMessage
        defaultMessage="Hidden"
        id="dapp.accounts.hidden"
        values={Object {}}
      />
    </Label>
  </CardContent>
</Card>
`;

exports[`should render correctly in when isDefault 1`] = `
<Card
  className=" "
  onClick={null}
>
  <CardContent>
    <Image
      address="123"
      as={[Function]}
      floated="right"
      size="mini"
      ui={true}
    />
    <CardHeader>
      <IdentityName
        address="123"
        name="Foo"
        unknown={true}
      />
    </CardHeader>
    <CardMeta>
      123
    </CardMeta>
    <CardDescription>
      <p>
        Bar
      </p>
      <p>
        <FormattedMessage
          defaultMessage="Balance"
          id="dapp.accounts.balance"
          values={Object {}}
        />
        : 
        <EtherBalance
          balance="Bond of 123"
        />
      </p>
    </CardDescription>
  </CardContent>
</Card>
`;
